### PR TITLE
unit testing dashboard.js and node.d plugins (groundworks)

### PR DESCRIPTION
**Pull request by BrainDoctor**
I assumed that the folder "netdata/tests" is a folder which doesn't interfere with the installers/upgrader and thus decided to create any test files in there.

I just had to. How can you even develop software without unit tests? ;P

Unit tests are to be expanded as needed. Not every aspect is covered (yet, especially jQuery). I'm curious on your thoughts. Also WebStorm is not necessary as all tests can be run from CLI.

Leaving the readme here:

# Installation

Tested on Linux Mint 18.2 Sara (Ubuntu/debian derivative)

Make sure you are the user who is developer (permissions, except sudo ofc)

```sh
sudo apt-get install nodejs npm chromium-browser
cd /path/to/your/netdata
npm install
cd node_modules
git clone https://github.com/BrainDoctor/jasmine-node.git
```

That should install the necessary node modules.

Other browsers work too (Chrome, Firefox). However, only the Chromium Browser 59 has been tested for headless unit testing.

## Versions

The commands above leave me with the following versions (July 2017):

 - nodejs: v4.2.6
 - npm: 3.5.2
 - chromium-browser: 59.0.3071.109
 - WebStorm (optional): 2017.1.4

# Configuration

## NPM

The dependencies are installed in `netdata/package.json`. If you install a new NPM module, it gets added here. Future developers just need to execute `npm install` and every dep gets added automatically.

## Karma

Karma configuration is in `tests/web/karma.conf.js`. Documentation is provided via comments.

## WebStorm

If you use the JetBrains WebStorm IDE, you can integrate the karma runtime.

### for Karma (Client side testing)

Headless Chromium:
1. Run > Edit Configurations
2. "+" > Karma
3. - Name: Karma Headless Chromium
   - Configuration file: /path/to/your/netdata/tests/web/karma.conf.js
   - Browsers to start: ChromiumHeadless
   - Node interpreter: /usr/bin/nodejs (MUST be absolute, NVM works too)
   - Karma package: /path/to/your/netdata/node_modules/karma

GUI Chromium is similar:
1. Run > Edit Configurations
2. "+" > Karma
3. - Name: Karma Chromium
   - Configuration file: /path/to/your/netdata/tests/web/karma.conf.js
   - Browsers to start: Chromium
   - Node interpreter: /usr/bin/nodejs (MUST be absolute, NVM works too)
   - Karma package: /path/to/your/netdata/node_modules/karma

You may add other browsers too (comma separated). With the "Browsers to start" field you can override any settings in karma.conf.js.

Also it is recommended to install WebStorm IDE Extension/Addon to Chrome/Chromium for awesome debugging.

### for node.d plugins (nodejs)

1. Run > Edit Configurations
2. "+" > Node.js
3. - Name: Node.d plugins
   - Node interpreter: /usr/bin/nodejs (MUST be absolute, NVM works too)
   - JavaScript file: node_modules/jasmine-node/bin/jasmine-node
   - Application parameters: --captureExceptions tests/node.d

# Running

### Karma
Just run the configured run configurations and they produce nice test trees:

karma_run_2

### node.js

Debugging is awesome too!
node_debug

### Karma

```sh
cd /path/to/your/netdata

nodejs ./node_modules/karma/bin/karma start tests/web/karma.conf.js --single-run=true --browsers=ChromiumHeadless
```
will start the karma server, start chromium in headless mode and exit.

If a test fails, it produces even a stack trace:
karma_run_1

### Node.d plugins

```sh
cd /path/to/your/netdata

nodejs node_modules/jasmine-node/bin/jasmine-node --captureExceptions tests/node.d
```

will run the tests in `tests/node.d` and produce a stacktrace too on error:
node_run

### Karma

A nice HTML is produced from Karma which shows which code paths were executed. It is located somewhere in `/path/to/your/netdata/coverage/`

coverage_2
and
coverage_1

### Node.d

Apparently, jasmine-node can produce a junit report with the `--junitreport` flag. But that output was not very useful. Maybe it's configurable?

## CI

The karma and node.d runners can be integrated in Travis (AFAIK), but that is outside my ability.

Note: Karma is for browser-testing. On a build server, no GUI or browser might by available, unless browsers support headless mode.

### CODE REVIEW

**Comment by ktsaou** [comment]
Wow! This is really nice!
Give me some time to review it. Each release puts some pressure on me, so give me a few days...

**Comment by BrainDoctor** [comment]
no problem, there's no rush ;)

**Comment by BrainDoctor** [comment]
I refactored the fronius plugin. Thanks to unit testing it still works the same, but I believe the quality is better now. There are now more, smaller methods that better describe what they are doing instead of commenting everything. Also there are already 50 tests so far, executed in 0.15s on a VM ;)

**Comment by ktsaou** [comment]
I am trying to set it up.

I get this:

```
$ npm install
/home/costa/src/netdata-BrainDoctor.git
├─┬ jasmine@2.6.0 
│ ├── exit@0.1.2 
│ └─┬ glob@7.1.2 
│   ├── fs.realpath@1.0.0 
│   ├─┬ inflight@1.0.6 
│   │ └── wrappy@1.0.2 
│   ├── inherits@2.0.3 
│   ├── once@1.4.0 
│   └── path-is-absolute@1.0.1 
├── jasmine-core@2.6.4 
├─┬ karma@1.7.0 
│ ├── bluebird@3.5.0 
│ ├─┬ body-parser@1.17.2 
│ │ ├── bytes@2.4.0 
│ │ ├── content-type@1.0.2 
│ │ ├─┬ debug@2.6.7 
│ │ │ └── ms@2.0.0 
│ │ ├── depd@1.1.0 
│ │ ├─┬ http-errors@1.6.1 
│ │ │ ├── setprototypeof@1.0.3 
│ │ │ └── statuses@1.3.1 
│ │ ├── iconv-lite@0.4.15 
│ │ ├─┬ on-finished@2.3.0 
│ │ │ └── ee-first@1.1.1 
│ │ ├── qs@6.4.0 
│ │ ├─┬ raw-body@2.2.0 
│ │ │ └── unpipe@1.0.0 
│ │ └─┬ type-is@1.6.15 
│ │   ├── media-typer@0.3.0 
│ │   └─┬ mime-types@2.1.15 
│ │     └── mime-db@1.27.0 
│ ├─┬ chokidar@1.7.0 
│ │ ├─┬ anymatch@1.3.0 
│ │ │ ├── arrify@1.0.1 
│ │ │ └─┬ micromatch@2.3.11 
│ │ │   ├─┬ arr-diff@2.0.0 
│ │ │   │ └── arr-flatten@1.1.0 
│ │ │   ├─┬ braces@1.8.5 
│ │ │   │ ├─┬ expand-range@1.8.2 
│ │ │   │ │ └─┬ fill-range@2.2.3 
│ │ │   │ │   ├── is-number@2.1.0 
│ │ │   │ │   ├── isobject@2.1.0 
│ │ │   │ │   ├─┬ randomatic@1.1.7 
│ │ │   │ │   │ ├─┬ is-number@3.0.0 
│ │ │   │ │   │ │ └── kind-of@3.2.2 
│ │ │   │ │   │ └── kind-of@4.0.0 
│ │ │   │ │   └── repeat-string@1.6.1 
│ │ │   │ ├── preserve@0.2.0 
│ │ │   │ └── repeat-element@1.1.2 
│ │ │   ├─┬ expand-brackets@0.1.5 
│ │ │   │ └── is-posix-bracket@0.1.1 
│ │ │   ├── extglob@0.3.2 
│ │ │   ├── filename-regex@2.0.1 
│ │ │   ├─┬ kind-of@3.2.2 
│ │ │   │ └── is-buffer@1.1.5 
│ │ │   ├─┬ normalize-path@2.1.1 
│ │ │   │ └── remove-trailing-separator@1.0.2 
│ │ │   ├─┬ object.omit@2.0.1 
│ │ │   │ ├─┬ for-own@0.1.5 
│ │ │   │ │ └── for-in@1.0.2 
│ │ │   │ └── is-extendable@0.1.1 
│ │ │   ├─┬ parse-glob@3.0.4 
│ │ │   │ ├── glob-base@0.3.0 
│ │ │   │ └── is-dotfile@1.0.3 
│ │ │   └─┬ regex-cache@0.4.3 
│ │ │     ├── is-equal-shallow@0.1.3 
│ │ │     └── is-primitive@2.0.0 
│ │ ├── async-each@1.0.1 
│ │ ├── glob-parent@2.0.0 
│ │ ├─┬ is-binary-path@1.0.1 
│ │ │ └── binary-extensions@1.9.0 
│ │ ├─┬ is-glob@2.0.1 
│ │ │ └── is-extglob@1.0.0 
│ │ └─┬ readdirp@2.1.0 
│ │   ├─┬ readable-stream@2.3.3 
│ │   │ ├── isarray@1.0.0 
│ │   │ ├── process-nextick-args@1.0.7 
│ │   │ ├── string_decoder@1.0.3 
│ │   │ └── util-deprecate@1.0.2 
│ │   └── set-immediate-shim@1.0.1 
│ ├── colors@1.1.2 
│ ├─┬ combine-lists@1.0.1 
│ │ └── lodash@4.17.4 
│ ├─┬ connect@3.6.2 
│ │ ├─┬ finalhandler@1.0.3 
│ │ │ ├── encodeurl@1.0.1 
│ │ │ └── escape-html@1.0.3 
│ │ ├── parseurl@1.3.1 
│ │ └── utils-merge@1.0.0 
│ ├── core-js@2.4.1 
│ ├── di@0.0.1 
│ ├─┬ dom-serialize@2.2.1 
│ │ ├── custom-event@1.0.1 
│ │ ├── ent@2.2.0 
│ │ ├── extend@3.0.1 
│ │ └── void-elements@2.0.1 
│ ├─┬ expand-braces@0.1.2 
│ │ ├── array-slice@0.2.3 
│ │ ├── array-unique@0.2.1 
│ │ └─┬ braces@0.1.5 
│ │   └─┬ expand-range@0.1.1 
│ │     ├── is-number@0.1.1 
│ │     └── repeat-string@0.2.2 
│ ├── graceful-fs@4.1.11 
│ ├─┬ http-proxy@1.16.2 
│ │ ├── eventemitter3@1.2.0 
│ │ └── requires-port@1.0.0 
│ ├── isbinaryfile@3.0.2 
│ ├── lodash@3.10.1 
│ ├─┬ log4js@0.6.38 
│ │ ├─┬ readable-stream@1.0.34 
│ │ │ ├── core-util-is@1.0.2 
│ │ │ ├── isarray@0.0.1 
│ │ │ └── string_decoder@0.10.31 
│ │ └── semver@4.3.6 
│ ├── mime@1.3.6 
│ ├─┬ minimatch@3.0.4 
│ │ └─┬ brace-expansion@1.1.8 
│ │   ├── balanced-match@1.0.0 
│ │   └── concat-map@0.0.1 
│ ├─┬ optimist@0.6.1 
│ │ ├── minimist@0.0.10 
│ │ └── wordwrap@0.0.3 
│ ├── qjobs@1.1.5 
│ ├── range-parser@1.2.0 
│ ├── rimraf@2.6.1 
│ ├── safe-buffer@5.1.1 
│ ├─┬ socket.io@1.7.3 
│ │ ├─┬ debug@2.3.3 
│ │ │ └── ms@0.7.2 
│ │ ├─┬ engine.io@1.8.3 
│ │ │ ├─┬ accepts@1.3.3 
│ │ │ │ └── negotiator@0.6.1 
│ │ │ ├── base64id@1.0.0 
│ │ │ ├── cookie@0.3.1 
│ │ │ ├─┬ debug@2.3.3 
│ │ │ │ └── ms@0.7.2 
│ │ │ ├─┬ engine.io-parser@1.3.2 
│ │ │ │ ├── after@0.8.2 
│ │ │ │ ├── arraybuffer.slice@0.0.6 
│ │ │ │ ├── base64-arraybuffer@0.1.5 
│ │ │ │ ├── blob@0.0.4 
│ │ │ │ └── wtf-8@1.0.0 
│ │ │ └─┬ ws@1.1.2 
│ │ │   ├── options@0.0.6 
│ │ │   └── ultron@1.0.2 
│ │ ├─┬ has-binary@0.1.7 
│ │ │ └── isarray@0.0.1 
│ │ ├── object-assign@4.1.0 
│ │ ├─┬ socket.io-adapter@0.5.0 
│ │ │ └─┬ debug@2.3.3 
│ │ │   └── ms@0.7.2 
│ │ ├─┬ socket.io-client@1.7.3 
│ │ │ ├── backo2@1.0.2 
│ │ │ ├── component-bind@1.0.0 
│ │ │ ├── component-emitter@1.2.1 
│ │ │ ├─┬ debug@2.3.3 
│ │ │ │ └── ms@0.7.2 
│ │ │ ├─┬ engine.io-client@1.8.3 
│ │ │ │ ├── component-emitter@1.2.1 
│ │ │ │ ├── component-inherit@0.0.3 
│ │ │ │ ├─┬ debug@2.3.3 
│ │ │ │ │ └── ms@0.7.2 
│ │ │ │ ├── has-cors@1.1.0 
│ │ │ │ ├── parsejson@0.0.3 
│ │ │ │ ├── parseqs@0.0.5 
│ │ │ │ ├── xmlhttprequest-ssl@1.5.3 
│ │ │ │ └── yeast@0.1.2 
│ │ │ ├── indexof@0.0.1 
│ │ │ ├── object-component@0.0.3 
│ │ │ ├─┬ parseuri@0.0.5 
│ │ │ │ └─┬ better-assert@1.0.2 
│ │ │ │   └── callsite@1.0.0 
│ │ │ └── to-array@0.1.4 
│ │ └─┬ socket.io-parser@2.3.1 
│ │   ├── component-emitter@1.1.2 
│ │   ├─┬ debug@2.2.0 
│ │   │ └── ms@0.7.1 
│ │   ├── isarray@0.0.1 
│ │   └── json3@3.3.2 
│ ├── source-map@0.5.6 
│ ├─┬ tmp@0.0.31 
│ │ └── os-tmpdir@1.0.2 
│ └─┬ useragent@2.2.1 
│   └── lru-cache@2.2.4 
├─┬ karma-chrome-launcher@2.2.0 
│ ├─┬ fs-access@1.0.1 
│ │ └── null-check@1.0.0 
│ └─┬ which@1.2.14 
│   └── isexe@2.0.0 
├─┬ karma-coverage@1.1.1 
│ ├─┬ dateformat@1.0.12 
│ │ ├── get-stdin@4.0.1 
│ │ └─┬ meow@3.7.0 
│ │   ├─┬ camelcase-keys@2.1.0 
│ │   │ └── camelcase@2.1.1 
│ │   ├── decamelize@1.2.0 
│ │   ├─┬ loud-rejection@1.6.0 
│ │   │ ├─┬ currently-unhandled@0.4.1 
│ │   │ │ └── array-find-index@1.0.2 
│ │   │ └── signal-exit@3.0.2 
│ │   ├── map-obj@1.0.1 
│ │   ├── minimist@1.2.0 
│ │   ├─┬ normalize-package-data@2.4.0 
│ │   │ ├── hosted-git-info@2.5.0 
│ │   │ ├─┬ is-builtin-module@1.0.0 
│ │   │ │ └── builtin-modules@1.1.1 
│ │   │ └─┬ validate-npm-package-license@3.0.1 
│ │   │   ├─┬ spdx-correct@1.0.2 
│ │   │   │ └── spdx-license-ids@1.2.2 
│ │   │   └── spdx-expression-parse@1.0.4 
│ │   ├─┬ read-pkg-up@1.0.1 
│ │   │ ├─┬ find-up@1.1.2 
│ │   │ │ ├── path-exists@2.1.0 
│ │   │ │ └─┬ pinkie-promise@2.0.1 
│ │   │ │   └── pinkie@2.0.4 
│ │   │ └─┬ read-pkg@1.1.0 
│ │   │   ├─┬ load-json-file@1.1.0 
│ │   │   │ ├─┬ parse-json@2.2.0 
│ │   │   │ │ └─┬ error-ex@1.3.1 
│ │   │   │ │   └── is-arrayish@0.2.1 
│ │   │   │ ├── pify@2.3.0 
│ │   │   │ └─┬ strip-bom@2.0.0 
│ │   │   │   └── is-utf8@0.2.1 
│ │   │   └── path-type@1.1.0 
│ │   ├─┬ redent@1.0.0 
│ │   │ ├─┬ indent-string@2.1.0 
│ │   │ │ └─┬ repeating@2.0.1 
│ │   │ │   └─┬ is-finite@1.0.2 
│ │   │ │     └── number-is-nan@1.0.1 
│ │   │ └── strip-indent@1.0.1 
│ │   └── trim-newlines@1.0.0 
│ └─┬ istanbul@0.4.5 
│   ├── abbrev@1.0.9 
│   ├── async@1.5.2 
│   ├─┬ escodegen@1.8.1 
│   │ ├── estraverse@1.9.3 
│   │ ├── esutils@2.0.2 
│   │ ├─┬ optionator@0.8.2 
│   │ │ ├── deep-is@0.1.3 
│   │ │ ├── fast-levenshtein@2.0.6 
│   │ │ ├── levn@0.3.0 
│   │ │ ├── prelude-ls@1.1.2 
│   │ │ ├── type-check@0.3.2 
│   │ │ └── wordwrap@1.0.0 
│   │ └─┬ source-map@0.2.0 
│   │   └── amdefine@1.0.1 
│   ├── esprima@2.7.3 
│   ├── glob@5.0.15 
│   ├─┬ handlebars@4.0.10 
│   │ ├── source-map@0.4.4 
│   │ └─┬ uglify-js@2.8.29 
│   │   ├── uglify-to-browserify@1.0.2 
│   │   └─┬ yargs@3.10.0 
│   │     ├── camelcase@1.2.1 
│   │     ├─┬ cliui@2.1.0 
│   │     │ ├─┬ center-align@0.1.3 
│   │     │ │ ├─┬ align-text@0.1.4 
│   │     │ │ │ └── longest@1.0.1 
│   │     │ │ └── lazy-cache@1.0.4 
│   │     │ ├── right-align@0.1.3 
│   │     │ └── wordwrap@0.0.2 
│   │     └── window-size@0.1.0 
│   ├─┬ js-yaml@3.9.0 
│   │ ├─┬ argparse@1.0.9 
│   │ │ └── sprintf-js@1.0.3 
│   │ └── esprima@4.0.0 
│   ├─┬ mkdirp@0.5.1 
│   │ └── minimist@0.0.8 
│   ├── nopt@3.0.6 
│   ├── resolve@1.1.7 
│   ├─┬ supports-color@3.2.3 
│   │ └── has-flag@1.0.0 
│   └── wordwrap@1.0.0 
├── karma-firefox-launcher@1.0.1 
└── karma-jasmine@1.1.0 

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN netdata-BrainDoctor.git No repository field.
npm WARN netdata-BrainDoctor.git No license field.
```

**Comment by ktsaou** [comment]
Also, it seems I don't have `jasmine-node`:

```sh
# find . -name \*jasmine\*
./tests/web/lib/jasmine-jquery.js
./node_modules/jasmine
./node_modules/jasmine/tasks/jasmine.js
./node_modules/jasmine/bin/jasmine.js
./node_modules/jasmine/lib/jasmine.js
./node_modules/jasmine/lib/examples/jasmine.json
./node_modules/karma-jasmine
./node_modules/.bin/jasmine
./node_modules/jasmine-core
./node_modules/jasmine-core/lib/jasmine-core.js
./node_modules/jasmine-core/lib/jasmine-core
./node_modules/jasmine-core/lib/jasmine-core/example/node_example/lib/jasmine_examples
./node_modules/jasmine-core/lib/jasmine-core/example/node_example/spec/jasmine_examples
./node_modules/jasmine-core/lib/jasmine-core/example/node_example/spec/helpers/jasmine_examples
./node_modules/jasmine-core/lib/jasmine-core/jasmine-html.js
./node_modules/jasmine-core/lib/jasmine-core/jasmine.css
./node_modules/jasmine-core/lib/jasmine-core/jasmine.js
./node_modules/jasmine-core/jasmine_core.egg-info
./node_modules/jasmine-core/images/jasmine_favicon.png
./node_modules/jasmine-core/images/jasmine-horizontal.svg
./node_modules/jasmine-core/images/jasmine-horizontal.png
```

**Comment by ktsaou** [comment]
(btw, I am excited about all these - thank you!)

**Comment by LeeNX** [comment]
@ktsaou - You one sick puppy, but that is Wicked!!! Thanks!

**Comment by BrainDoctor** [comment]
I stated in the readme that you need to git clone https://github.com/BrainDoctor/jasmine-node.git into node_modules. I had to make my own jasmine-node since the old was outdated.

**Comment by BrainDoctor** [comment]
I'll do a clean install to verify that I didn't miss anything :)

**Comment by BrainDoctor** [comment]
Alright, the package.json was missing some dependencies for jasmine-node. first, do a `sudo rm -r node_modules/jasmine-node` and then `npm install`. Then `cd node_modules` and `git clone https://github.com/BrainDoctor/jasmine-node.git` again. We have to remove the old because npm would complain otherwise.

**Comment by ktsaou** [comment]
Thanks! I installed jasmine-node and I get this when I attempt to run node.d modules tests:

```
/usr/bin/node /home/costa/src/netdata-BrainDoctor.git/node_modules/jasmine-node/bin/jasmine-node --captureExceptions tests/node.d
module.js:471
    throw err;
    ^

Error: Cannot find module 'coffee-script/register'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/costa/src/netdata-BrainDoctor.git/node_modules/jasmine-node/lib/jasmine-node/runner.js:13:12)
    at Object.<anonymous> (/home/costa/src/netdata-BrainDoctor.git/node_modules/jasmine-node/lib/jasmine-node/runner.js:212:4)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)

Process finished with exit code 1
```

**Comment by ktsaou** [comment]
Just saw your latest message.

`npm install` gives me this:

```
$ npm install
npm WARN deprecated minimatch@0.2.14: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated graceful-fs@1.2.3: graceful-fs v3.0.0 and before will fail on node releases >= v7.0. Please update to graceful-fs@^4.0.0 as soon as possible. Use 'npm ls graceful-fs' to find it in the tree.
minimist@0.0.8 node_modules/mkdirp/node_modules/minimist -> node_modules/istanbul/node_modules/minimist
/home/costa/src/netdata-BrainDoctor.git
├── coffee-script@1.12.7 
├─┬ gaze@0.5.2 
│ └─┬ globule@0.1.0 
│   ├─┬ glob@3.1.21 
│   │ ├── graceful-fs@1.2.3 
│   │ └── inherits@1.0.2 
│   ├── lodash@1.0.2 
│   └─┬ minimatch@0.2.14 
│     └── sigmund@1.0.1 
├── UNMET PEER DEPENDENCY grunt@>=0.4
├── grunt-exec@0.4.7 
├─┬ jasmine-growl-reporter@0.2.1 
│ └── growl@1.7.0 
├── jasmine-reporters@2.0.8 
├─┬ karma-coverage@1.1.1
│ └─┬ istanbul@0.4.5
│   └── mkdirp@0.5.1 
├── mkdirp@0.3.5 
├── underscore@1.6.0 
├── walkdir@0.0.11 
└─┬ xml2js@0.4.17 
  ├── sax@1.2.4 
  └─┬ xmlbuilder@4.2.1 
    └── lodash@4.17.4 

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN grunt-exec@0.4.7 requires a peer of grunt@>=0.4 but none was installed.
npm WARN netdata-BrainDoctor.git No repository field.
npm WARN netdata-BrainDoctor.git No license field.
```

I checked out again `jasmine-node` and I now it seems to work:

```
/usr/bin/node /home/costa/src/netdata-BrainDoctor.git/node_modules/jasmine-node/bin/jasmine-node --captureExceptions tests/node.d
..................................................

Finished in 0.103 seconds
50 Tests, 0 Failures, 0 Skipped


Process finished with exit code 0
```

**Comment by BrainDoctor** [comment]
yeah I have the same warnings, I just ignored them^^

congratulations, you just ran 50 unit tests ;) have you tested karma/chrome as well?

**Comment by ktsaou** [comment]
Yes I did. I need to get familiar with all these though...

I guess you have left the failed chrome test in purpose. Right?

As I see it, this kind of unit testing is extremely useful when you need to make sure something works, but it seems to be quite an overhead if the code base is not stable (i.e. you refactor it from time to time). Am I right?

File: tests/README.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+Make sure you are the user who is developer (permissions, except sudo ofc)
+
+```sh
+sudo apt-get install nodejs npm chromium-browser
+
+cd /path/to/your/netdata
+npm install
+cd node_modules
+git clone https://github.com/BrainDoctor/jasmine-node.git
```
Comment (simonnagl) [inline tests/README.md]: This dependency can also be added to package.json.

Check the documentation https://docs.npmjs.com/files/package.json#git-urls-as-dependencies

**Comment by simonnagl** [comment]
@BrainDoctor Great Work! I think this brings the netdata client really forward.

It would also nice to add the tests to the travis-ci pipeline. But maybe in another pull request...

File: tests/README.md (hunk 1 above)
Comment (BrainDoctor) [inline tests/README.md]: haha, I didn't know this^^ of course this is an even better approach.

**Comment by BrainDoctor** [comment]
@simonnagl yes, if the dependencies can be installed on the build server. And yes, this should be done in another PR and I'm not familiar with travis

@ktsaou Yes, that was on purpose to see whether the results of a failure produced some kind of report (a stacktrace or mentioning if assertions didn't work)

Now to the unit test overhead. Well yes, it creates a slight overhead, since the test code should always be up to date with the requirements. However, in the long term, writing and testing with unit tests SAVES time. Think of the problem in #2455. After discussing what should and shouldn't be (which are basically the requirements), we have to test the implementation. Now without unit tests, I need some sort of dashboard and charts to set up and verify the changes and confirm that they meet the requirements. But this is manually and takes a lot time. It's basically an integration or even system test (if you know the V model in software development).

With unit tests, I can quickly create a new test case and verify if the actual output matches with the expected output based on a specific input. Also, as long as the test cases are valid, they reduce the risk of collateral damage if a change might break anything or re-introduce an old but fixed bug -> In just 0.1 seconds you have an answer whether your changed broke a specification/test case. If thats the case you can investigate, but it probably saves you a new issue on github, which would require you to look at your code again and search for the problem. Before you pushed the change at all.

Now, I'm not saying that we should unit test ALL of the dashboard and other plugins. However, I believe that a gradual buildup of unit testing is worth the time. It could be done whenever a change or bugfix is required. If it's easy testable (e.g. a method just like the easypiechart calculation) then you are faster to write the tests than you can setup some test dashboard and charts.

Also, not everything is unit testable. As I stated, I could't figure out (yet) how to test code that is asynchronous or requires jquery or uses html source. I'm sure test doubles are possible though. Visual (GUI) or integration testing is still needed, since you can't know whether this color looks good on a specific background beforehand. But every automated test helps and is faster. Refactoring happens all the time anyway. It's also a good feeling to see that everything is green before pushing it :)

**Comment by ktsaou** [comment]
Thanks! Yes, I know unit testing in general. I use such tests at the core of netdata. Run:

```sh
sudo netdata -W unittest
```

Without it, it would be impossible to check all the cases of interpolation, database operations, data collection scenarios, etc every time I touch that part of the code. That part is quite complex and extremely time consuming to test, almost impossible to do it in reasonable time (because the test cases have to be prepared with great accuracy).

But ,this is the first time I saw UI and node.js unit testing and of course all the tools you used.

**Comment by BrainDoctor** [comment]
Oh! Then I wrote too much text ;D

In the newest commits you'll see some unit tests for easypiechart. I included your recent merge already and wrote some tests for it. And this one
```javascript
    it("should return 0.1 if value is zero and min/max undefined", function () {
        var state = createState(null, null);

        var result = NETDATA.easypiechartPercentFromValueMinMax(state, 0, -1, 2);

        expect(result).toBe(0.1);
    });
```
fails with with `Expected 0 to be 0.1.`

I'm assuming that always a non-zero value should be returned, is that correct?

**Comment by ktsaou** [comment]
> I'm assuming that always a non-zero value should be returned, is that correct?

No. The value of zero, does not show any color on the chart. It is just the grey background. So, if the range is positive I use 0.01 and when the range is negative -0.01. This makes the chart idle with a small bullet.

The only way to get zero back is when the range goes from negative to positive. In this case, the little idle bullet should be at the center, but it is not supported. It will either be a little bit to left or a little bit to the right. So, only in this case, I preferred to let disappear...

**Comment by ktsaou** [comment]
In gauge.js there is another bug: if the value is the absolute minimum or the absolute maximum, the chart is broken. So there I use 0.001% to 99.999%.

**Comment by BrainDoctor** [comment]
thanks, this clarified things up.

**Comment by BrainDoctor** [comment]
How would you like to proceed with this PR? I intend to refactor the stiebel eltron plugin ( which is still open in #2448 ) and add some unit tests, as I have done with fronius within this PR (I couldn't resist and shows some real-world test examples). I can't give a time estimation though. I guess it doesn't make much sense to review it when a refactor is underway anyway. If this PR is merged before 2448 and before I begin with the refactor, I can include the master branch it into PR #2448 before merging it, so you'll get to review the refactored plugin with its unit tests in the first time.

**Comment by ktsaou** [comment]
Well, I think we can merge both of them.

The only help I would like, is to add all the files to the Makefiles so that the release will not complain (it complains if the git includes files that are not defined in `Makefile.am` somewhere). Of course these files should not be installed, but still they need to marked as not to be installed.

Once we have this, I believe we can merge them.

**Comment by BrainDoctor** [comment]
I am very unfamiliar with makefile. Can't we just ignore the tests/* directory? Where exactly would these entries have to go?

**Comment by ktsaou** [comment]
ok, I did all the preparations at PR #2513 

You will have to rebase this PR to get the latest files and then edit `tests/Makefile.am` and add the files, like this:

```
nodist_noinst_DATA = \
    $(NULL)

nodist_noinst_SCRIPTS = \
	stress.sh \
    $(NULL)
```

In DATA you put everything except executable scripts which you should add them to SCRIPTS.
All files relative to `tests/`.

You should also add `package.json` at the root `Makefile.am` in section `dist_noinst_DATA`.

Please do the same for the plugin. I have added `fronius` to `node.d/Makefile.am`, so just do the same.

**Comment by ktsaou** [comment]
btw, `dist` means distribute with the package.
`noinst` means do not install when netdata is installed.

I'll merge #2513 in a while. Then you can rebase.

**Comment by ktsaou** [comment]
merged PR #2513 . Is it ok now? (if not, it is ok, I'll merge it and then fix the files).

**Comment by BrainDoctor** [comment]
ok, that should cover it.

**Comment by ktsaou** [comment]
I am ready to merge this.
Do you agree?

**Comment by BrainDoctor** [comment]
yes

**Comment by ktsaou** [comment]
merged! thank you!

**Pull request merged**